src/lib/parallel_scan/parallel_scan.ml

```diff
@@ -262,12 +225,31 @@ module State = struct
     | J.Merge (Some a, Some b) -> Some (A.Merge (a, b))
     | _ -> None
 
-  let jobs_ready state =
-    List.filter_map (read_jobs state) ~f:(fun (job, _) -> job_ready job)
+  let job state index =
+    match job_ready (Ring_buffer.read_i state.jobs index) with
+    | None ->
+        Or_error.errorf
```

Comment:
Can you explain why you return an error here now instead of short-circuiting (stopping the list we were creating) like we did before?

We assume snark work will be done most of the time, but I think it's possible to get into the case where the work is not done. Is that case handled different with the error now?

Reply:
ok so this is only if the data is bad, should we use an exception instead then? As long as we can verify the correctness of this queue when we deserialize the thing from the network?

Reply:
Well, I made it an error instead of an exception so that `Staged_ledger` or even the `Proposer` or `Transition_frontier` could handle it differently (maybe request the scan_state from the network or construct a new one from the previous diff or something else). For now, throwing an exception in `Ledger_builder`